### PR TITLE
Prevent from removing taxon that is in use by a promotion rule

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13                   |
| Bug fix?        | yes                                                       |
| New feature?    | no                                                      |
| BC breaks?      | no                                                      |
| Deprecations?   | no  |
| License         | MIT                                                          |

related to https://github.com/Sylius/Sylius/pull/15277

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/spec/EventSubscriber/TaxonDeletionEventSubscriberSpec.php [hunk 1]
```diff
@@ -81,4 +83,44 @@ function it_throws_an_exception_if_a_subject_is_menu_taxon(
...
+        $request->getMethod()->willReturn(Request::METHOD_POST);
+
+        $event = new ViewEvent(
+            $kernel->getWrappedObject(),
+            $request->getWrappedObject(),
+            HttpKernelInterface::MAIN_REQUEST,
+            $taxon->getWrappedObject(),
+        );
+
+        $taxonInPromotionRuleChecker->isInUse($taxon)->willReturn(false);
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/spec/EventSubscriber/TaxonDeletionEventSubscriberSpec.php:124]:
```suggestion
        $taxonInPromotionRuleChecker->isInUse($taxon)->shouldNotBeCalled();
```

File: src/Sylius/Bundle/ApiBundle/spec/EventSubscriber/TaxonDeletionEventSubscriberSpec.php [hunk 2]
```diff
@@ -81,4 +83,44 @@ function it_throws_an_exception_if_a_subject_is_menu_taxon(
...
+
+        $this->shouldNotThrow()->during('protectFromRemovingTaxonInUseByPromotionRule', [$event]);
+    }
+
+    function it_throws_an_exception_when_trying_to_delete_taxon_assigned_to_promotion_rule(
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/spec/EventSubscriber/TaxonDeletionEventSubscriberSpec.php]: Missing the case when it's being deleted but it's not in use

File: src/Sylius/Bundle/ApiBundle/spec/EventSubscriber/TaxonDeletionEventSubscriberSpec.php [hunk 3]
```diff
@@ -81,4 +83,44 @@ function it_throws_an_exception_if_a_subject_is_menu_taxon(
...
+        $event = new ViewEvent(
+            $kernel->getWrappedObject(),
+            $request->getWrappedObject(),
+            HttpKernelInterface::MAIN_REQUEST,
+            $taxon->getWrappedObject(),
+        );
+
+        $taxonInPromotionRuleChecker->isInUse($taxon)->willReturn(true);
+
+        $this->shouldThrow(TaxonCannotBeRemoved::class)->during('protectFromRemovingTaxonInUseByPromotionRule', [$event]);
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/spec/EventSubscriber/TaxonDeletionEventSubscriberSpec.php]:
```suggestion
        $this
            ->shouldThrow(TaxonCannotBeRemoved::class)
            ->during('protectFromRemovingTaxonInUseByPromotionRule', [$event])
        ;
```

File: src/Sylius/Bundle/ApiBundle/spec/EventSubscriber/TaxonDeletionEventSubscriberSpec.php [hunk 4]
```diff
@@ -14,18 +14,20 @@
...
 use Sylius\Component\Core\Model\TaxonInterface;
+use Sylius\Component\Core\Promotion\Checker\TaxonInPromotionRuleCheckerInterface;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpKernel\Event\ViewEvent;
 use Symfony\Component\HttpKernel\HttpKernelInterface;
 
 final class TaxonDeletionEventSubscriberSpec extends ObjectBehavior
 {
-    function let(ChannelRepositoryInterface $channelRepository): void
+    function let(ChannelRepositoryInterface $channelRepository, TaxonInPromotionRuleCheckerInterface $taxonInPromotionRuleChecker): void
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/spec/EventSubscriber/TaxonDeletionEventSubscriberSpec.php]:
```suggestion
    function let(
        ChannelRepositoryInterface $channelRepository,
        TaxonInPromotionRuleCheckerInterface $taxonInPromotionRuleChecker,
    ): void {
```

File: src/Sylius/Bundle/ApiBundle/Exception/TaxonCannotBeRemoved.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Exception;
+
+/** @experimental */
+final class TaxonCannotBeRemoved extends \RuntimeException
+{
+    public function __construct(string $message = 'Cannot delete, the taxon is in use.')
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Exception/TaxonCannotBeRemoved.php]: If this PR will be merged first https://github.com/Sylius/Sylius/pull/15322/files
please synchronize updated constructor signature.

**Review by jakubtobiasz — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Exception/TaxonCannotBeRemoved.php (hunk 5 above)
Comment (jakubtobiasz) [reply src/Sylius/Bundle/ApiBundle/Exception/TaxonCannotBeRemoved.php]: #15322 is merged :v

**Review by Rafikooo — APPROVED** [review]

**Review by NoResponseMate — APPROVED** [review]

**Comment by jakubtobiasz** [comment]
Thank you, @TheMilek!

**Pull request merged**